### PR TITLE
Added ZLIB_REGISTER to disable register keyword for C++.

### PR DESCRIPTION
**Pull request by nmoinvaz**

### CODE REVIEW

**Comment by Dead2** [comment]
Looks really good.
I am a little uncertain about the `ZLIB_REGISTER` name though. Would `ZREGISTER` or `Z_REGISTER` be better?
I know this is nitpicking, but any thoughts?

**Comment by nmoinvaz** [comment]
I am ok with how it is. But I think if we change it we should also change `ZLIB_INTERNAL`.

**Comment by nmoinvaz** [comment]
Maybe `ZREGISTER` and `ZINTERNAL` to match `ZSWAP` type stuff also in `zutil.h`.

**Comment by dankegel** [comment]
Z is too generic a prefix, could clash with other things...?

**Comment by mtl1979** [comment]
We should use "ZNG" prefix for all zlib-ng specific macros... Just to be consistent...

**Comment by nmoinvaz** [comment]
I was going to do something like this to preserve compatibility with anybody using -DZLIB_INTERNAL:
```
#ifdef ZLIB_INTERNAL
#  define ZINTERNAL ZLIB_INTERNAL
#else
#  define ZINTERNAL
#endif
#ifndef __cplusplus
#  define ZREGISTER register
#else
#  define ZREGISTER
#endif
```

But I think renaming can probably wait for another PR.

**Comment by Dead2** [comment]
ZNG_ should be pretty unique

**Comment by nmoinvaz** [comment]
I wouldn't think `ZREGISTER/ZINTERNAL` would conflict with anything either, since they are only used internally. @Dead2 are you saying you want me to rename ZLIB_INTERNAL to ZNG_INTERNAL? Why isn't `PREFIX` named `ZNG_PREFIX`?

**Comment by Dead2** [comment]
You are right, these should not leak out of our codebase.
So, ZREGISTER/ZINTERNAL should be perfectly fine (and shorter).

For defines that are externally visible however, we should maybe consider uniqueness.

File: zlib.h [hunk 1]
```diff
@@ -1782,7 +1782,7 @@ ZEXTERN int ZEXPORT gzgetc_(gzFile file);  /* backward compatibility */
    ZEXTERN void ZEXPORT crc32_combine_gen64(uint32_t *op, z_off64_t);
 #endif
 
-#if !defined(ZLIB_INTERNAL) && defined(Z_WANT64)
+#if !defined(ZINTERNAL) && defined(Z_WANT64)
```
Comment (nmoinvaz) [inline zlib.h]: I am unsure about this change here. It is in zlib.h. Perhaps we should leave this one to `ZLIB_INTERNAL`.

File: zconf-ng.h.in [hunk 2]
```diff
@@ -59,7 +59,7 @@
  * This is not mandatory, but it offers a little performance increase.
  */
 #if defined(ZLIB_DLL) && (defined(_WIN32) || (__has_declspec_attribute(dllexport) && __has_declspec_attribute(dllimport)))
-#  ifdef ZLIB_INTERNAL
+#  ifdef ZINTERNAL
```
Comment (nmoinvaz) [inline zconf-ng.h.in]: Unsure about this one too since it is in `zconf-ng.h`.

File: zconf.h.in [hunk 3]
```diff
@@ -63,7 +63,7 @@
  * This is not mandatory, but it offers a little performance increase.
  */
 #if defined(ZLIB_DLL) && (defined(_WIN32) || (__has_declspec_attribute(dllexport) && __has_declspec_attribute(dllimport)))
-#  ifdef ZLIB_INTERNAL
+#  ifdef ZINTERNAL
```
Comment (nmoinvaz) [inline zconf.h.in]: Unsure about this one too since it is in `zconf.h`.

File: zlib.h (hunk 1 above)
Comment (Dead2) [reply zlib.h]: This basically checks whether we are currently compiling zlib(-ng) or compiling an external application. So the define itself should never come from an external source. Therefore I think this is fine.

File: zconf.h.in (hunk 3 above)
Comment (Dead2) [reply zconf.h.in]: Uncertain, but I suspect same as zlib.h one.

File: zconf-ng.h.in (hunk 2 above)
Comment (Dead2) [reply zconf-ng.h.in]: Same as the one in zlib.h, so I think this is OK.

**Comment by Dead2** [comment]
Looks good to me.

**Comment by dankegel** [comment]
I disagree.  If the symbol is referenced in an installed .h file, it's public, and exposed to defines from other libraries.  
It might be ok if it were "unique enough", but my rule of thumb for "unique enough" is "no google hits", and it fails that test.
It doesn't help that it's in zconf.h rather than zlib.h, since zlib.h includes zconf.h.

It would be safer to keep it as ZLIB_INTERNAL.

IMHO a change to fix a problem related to the use of the register keyword should not change the symbols in the public .h file.  If you want to change ZLIB_INTERNAL, make that a separate pull request.

**Comment by Dead2** [comment]
@dankegel You are right of course, I only considered that we would only define that in our internal use and that would never affect an external program, but the check would of course also be affected if an external program sets it.

So, yeah. Keep ZLIB_INTERNAL as-is, and name only ZREGISTER?
Alternatively we could set ZLIB_INTERNAL and only use that in the externally visible header files, and shorten the one we use a lot internally. But that is better for a separate PR, yes.

**Comment by nmoinvaz** [comment]
I agree, I think it should be limited in scope. I have removed the second commit, so this one should only be related to `register` changes. I will do a separate PR after this one hits for the other changes. Not sure _discussion_ label is needed anymore.

**Comment by nmoinvaz** [comment]
Rebased.

**Pull request merged**